packages/mui-base/src/ListboxUnstyled/useListbox.ts

```diff
@@ -254,71 +242,96 @@ export default function useListbox<TOption>(props: UseListboxParameters<TOption>
-      );
-    } else {
-      selected = optionComparer(option, selectedValue as TOption);
-    }
-
-    const disabled = isOptionDisabled(option, index);
+  const getOptionState = React.useCallback(
+    (option: TOption): OptionState => {
+      let selected: boolean;
+      const index = options.findIndex((opt) => optionComparer(opt, option));
+      if (multiple) {
+        selected = ((latestSelectedValue.current as TOption[]) ?? []).some(
+          (value) => value != null && optionComparer(option, value),
+        );
+      } else {
+        selected = optionComparer(option, latestSelectedValue.current as TOption);
+      }
 
-    return {
-      selected,
-      disabled,
-      highlighted: highlightedIndex === index && index !== -1,
-    };
-  };
+      const disabled = isOptionDisabled(option, index);
+      const highlighted = latestHighlightedIndex.current === index && index !== -1;
 
-  const getOptionTabIndex = (optionState: OptionState) => {
-    if (focusManagement === 'activeDescendant') {
-      return undefined;
-    }
+      return {
+        selected,
+        disabled,
+        highlighted,
+      };
+    },
+    [
+      options,
+      multiple,
+      isOptionDisabled,
+      optionComparer,
+      latestSelectedValue,
+      latestHighlightedIndex,
+    ],
+  );
 
-    if (!optionState.highlighted) {
-      return -1;
-    }
+  const getOptionTabIndex = React.useCallback(
+    (optionState: OptionState) => {
+      if (focusManagement === 'activeDescendant') {
+        return undefined;
+      }
 
-    if (optionState.disabled && !disabledItemsFocusable) {
-      return -1;
-    }
+      if (!optionState.highlighted) {
+        return -1;
+      }
 
-    return 0;
-  };
+      if (optionState.disabled && !disabledItemsFocusable) {
+        return -1;
+      }
 
-  const getOptionProps = <TOther extends EventHandlers = {}>(
-    option: TOption,
-    otherHandlers: TOther = {} as TOther,
-  ): UseListboxOptionSlotProps<TOther> => {
-    const optionState = getOptionState(option);
-    const index = options.findIndex((opt) => optionComparer(opt, option));
+      return 0;
+    },
+    [focusManagement, disabledItemsFocusable],
+  );
 
-    return {
-      ...otherHandlers,
-      'aria-disabled': optionState.disabled || undefined,
-      'aria-selected': optionState.selected,
-      id: optionIdGenerator(option, index),
-      onClick: createHandleOptionClick(option, otherHandlers),
-      onPointerOver: createHandleOptionPointerOver(option, otherHandlers),
-      role: 'option',
-      tabIndex: getOptionTabIndex(optionState),
-    };
-  };
+  const getOptionProps = React.useCallback(
+    <TOther extends EventHandlers = {}>(
+      option: TOption,
+      otherHandlers: TOther = {} as TOther,
+    ): UseListboxOptionSlotProps<TOther> => {
+      const optionState = getOptionState(option);
+      const index = options.findIndex((opt) => optionComparer(opt, option));
```

Comment:
```suggestion
      const optionState = getOptionState(option);
      const { index } = optionState;
```

nit: we already are finding the index in getOptionState, we could return it as part of the state. Not sure if it's worth the change

Reply:
Good point! This could make it even faster a bit.